File: src/notification.rs
use chrono::DateTime;
use eyre::Context;
use eyre::Result;
use serde::{Deserialize, Serialize};
use std::env;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct State {
    pub last_updated_at: Option<String>,
}

impl State {
    pub fn try_create() -> Result<Self> {
        let path = env::current_exe()
            .unwrap()
            .parent()
            .unwrap()
            .join("state.json");
        if !path.exists() {
            log::debug!("state file not exists at {}", path.display());
            return Ok(Self {
                last_updated_at: None,
            });
        }
        let content = std::fs::read_to_string(&path).context("Read")?;
        let state: Self = serde_json::from_str(&content)?;
        Ok(state)
    }
    pub fn save(&mut self) {
        let path = env::current_exe()
            .unwrap()
            .parent()
            .unwrap()
            .join("state.json");
        log::debug!("save state {}", path.display());
        std::fs::write(path, serde_json::to_string_pretty(&self).unwrap()).unwrap();
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Notification {
    pub id: i128,
    pub kind: String,
    pub title: String,
    pub url: Option<String>,
    pub repo_url: Option<String>,
    pub repo_name: Option<String>,
    pub updated_at: Option<String>,
}

impl Notification {
    pub fn is_new(&self, state: &mut State) -> bool {
        // If new notification different from previous it's new
        if let (Ok(state_updated_at), Ok(updated_at)) = (
            DateTime::parse_from_rfc3339(&state.last_updated_at.clone().unwrap_or_default()),
            DateTime::parse_from_rfc3339(&self.updated_at.clone().unwrap_or_default()),
        ) {
            // Compare the timestamps
            return updated_at > state_updated_at;
        }
        true
    }
}

File: src/notification.py
import os
import json
from datetime import datetime
from typing import Optional
from log import log
from pathlib import Path

STATE_PATH = os.getenv('STATE_PATH') or 'state.json'
STATE_PATH = Path(STATE_PATH)

class State:
    def __init__(self, last_updated_at: Optional[str] = None):
        self.last_updated_at = last_updated_at

    @classmethod
    def try_create(cls):
        if not STATE_PATH.exists():
            log.info(f"Create empty state at {STATE_PATH}")
            with open(STATE_PATH, 'w') as f:
                f.write(json.dumps({}))
            return cls(last_updated_at=None)
        with open(STATE_PATH, 'r') as f:
            content = f.read()
        state: dict = json.loads(content)
        return cls(last_updated_at=state.get('last_updated_at'))

    def save(self):
        log.info(f"Saving state to {STATE_PATH}")
        with open(STATE_PATH, 'w') as f:
            json.dump(self.__dict__, f, indent=4)

class Notification:
    def __init__(self, id: int, kind: str, title: str, url: Optional[str], repo_url: Optional[str], repo_name: Optional[str], updated_at: Optional[str]):
        self.id = id
        self.kind = kind
        self.title = title
        self.url = url
        self.repo_url = repo_url
        self.repo_name = repo_name
        self.updated_at = updated_at

    def is_new(self, state: State) -> bool:
        # If the new notification has a different timestamp from the previous one, it's new
        if state.last_updated_at and self.updated_at:
            try:
                state_updated_at = datetime.fromisoformat(state.last_updated_at)
                updated_at = datetime.fromisoformat(self.updated_at)
                return updated_at > state_updated_at
            except ValueError:
                return True
        return True
